### PR TITLE
repair/streaming: enable toggling tombstone gc with a config item

### PR DESCRIPTION
**Pull request by denesb**
We currently disable tombstone GC for compaction done on the read path of streaming and repair, because those expired tombstones can still prevent data resurrection. With time-based tombstone GC, missing a repair for long enough can cause data resurrection because a tombstone is potentially GC'd before it could be spread to every node by repair. So repair disseminating these expired tombstones helps clusters which missed repair for long enough. It is not a guarantee because compaction could have done the GC itself, but it is better than nothing.
This last resort is getting less important with repair-based tombstone GC. Furthermore, we have seen this cause huge repair amplification in a cluster, where expired tombstones triggered repair replicating otherwise identical rows.

This series makes tombstone GC on the streaming/repair compaction path configurable with a config item. This new config item defaults to `false` (current behaviour), setting it to `true`, will enable tombstone GC.

Fixes: https://github.com/scylladb/scylladb/issues/19015

Not a regression, no backport needed

### CODE REVIEW

**Comment by denesb** [comment]
@pdbossman @vladzcloudius

**Comment by denesb** [comment]
Conflicts with/depends on https://github.com/scylladb/scylladb/pull/18705.

**Comment by denesb** [comment]
Don't merge it yet, I also want to write a test.

**Comment by denesb** [comment]
Rebased onto master (https://github.com/scylladb/scylladb/pull/18705 was merged).

**Comment by denesb** [comment]
> ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9639/)

https://github.com/scylladb/scylladb/issues/19034 again.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping.

**Comment by avikivity** [comment]
L :eyes: king

**Comment by avikivity** [comment]
> This series makes tombstone GC on the streaming/repair compaction path configurable with a config item. This new config item defaults to false (current behaviour), setting it to true, will enable tombstone GC.

Should we make it true when tombstone_gc = repair?

File: db/config.cc [hunk 1]
```diff
@@ -970,6 +970,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
...
     , allowed_repair_based_node_ops(this, "allowed_repair_based_node_ops", liveness::LiveUpdate, value_status::Used, "replace,removenode,rebuild,bootstrap,decommission", "A comma separated list of node operations which are allowed to enable repair based node operations. The operations can be bootstrap, replace, removenode, decommission and rebuild.")
     , enable_compacting_data_for_streaming_and_repair(this, "enable_compacting_data_for_streaming_and_repair", liveness::LiveUpdate, value_status::Used, true, "Enable the compacting reader, which compacts the data for streaming and repair (load'n'stream included) before sending it to, or synchronizing it with peers. Can reduce the amount of data to be processed by removing dead data, but adds CPU overhead.")
+    , enable_tombstone_gc_for_streaming_and_repair_compaction(this, "enable_tombstone_gc_for_streaming_and_repair_compaction", liveness::LiveUpdate, value_status::Used, false,
+            "Enable compaction during streaming and repair (see enable_compacting_data_for_streaming_and_repair) to garbage-collect tombstones")
```
Comment (nyh) [inline db/config.cc:974]: I think that such an esoteric flag requires a whole paragraph to explain it, like you did in this cover letter - this one-sentence explanation won't mean anything to anyone a year from now. Perhaps it deserves a long explanation in https://opensource.docs.scylladb.com/stable/reference/configuration-parameters.html or some repair-related document.

Comment (avikivity) [inline db/config.cc:974]: The word "compaction" is misleading here. People will see it as a compaction process that reads some sstables and writes others. But here it's just about eliminating tombstones (and compacting_reader is just half of what users see as compaction.

File: test/topology_custom/test_repair.py [hunk 2]
```diff
@@ -67,3 +70,77 @@ async def test_enable_compacting_data_for_streaming_and_repair_live_update(manag
...
+    * enable_tombstone_gc_for_streaming_and_repair_compaction=0 works as expected
+    * enable_tombstone_gc_for_streaming_and_repair_compaction is live-update
+    """
+    cmdline = [
+            "--enable-compacting-data-for-streaming-and-repair", "1",
+            "--enable-tombstone-gc-for-streaming-and-repair-compaction", "1",
+            "--enable-cache", "0",
+            "--hinted-handoff-enabled", "0",
+            "--smp", "1",
+            "--logger-log-level", "api=trace:database=trace"]
```
Comment (nyh) [inline test/topology_custom/test_repair.py:90]: Are all these extra parameters (except the hinted-handoff) relevant to this test?

File: test/topology_custom/test_repair.py [hunk 3]
```diff
@@ -67,3 +70,77 @@ async def test_enable_compacting_data_for_streaming_and_repair_live_update(manag
...
+
+    cql.execute("CREATE KEYSPACE ks WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 2}")
+    cql.execute("CREATE TABLE ks.tbl (pk int, ck int, PRIMARY KEY (pk, ck)) WITH compaction = {'class': 'NullCompactionStrategy'}")
+
+    await manager.server_stop_gracefully(node2.server_id)
+
+    stmt = SimpleStatement("DELETE FROM ks.tbl WHERE pk = 0 AND ck = 0", consistency_level=ConsistencyLevel.ONE)
+    cql.execute(stmt)
+
+    await manager.server_start(node2.server_id, wait_others=1)
```
Comment (nyh) [inline test/topology_custom/test_repair.py:104]: We really need that feature to write to a single server and stop with those super-slow tests that start and stop servers just to exercise repair :-(

File: test/topology_custom/test_repair.py [hunk 4]
```diff
@@ -67,3 +70,77 @@ async def test_enable_compacting_data_for_streaming_and_repair_live_update(manag
...
+
+    # Flush memtables and remove commitlog, so we can freely GC tombstones.
+    await manager.server_restart(node1.server_id, wait_others=1)
```
Comment (nyh) [inline test/topology_custom/test_repair.py:107]: Nitpick: don't we have a REST API for flushing memtables instead of restarting the node? :-(

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
I'm ready to merge this and CI passed, but @avikivity asked for @asias to review so I'll wait.

File: db/config.cc (hunk 1 above)
Comment (asias) [reply db/config.cc:974]: We need to mention that setting enable_tombstone_gc_for_streaming_and_repair_compaction alone is not enough. It works only when nable_compacting_data_for_streaming_and_repair is set to true.

I am wondering if it is better to combine both options. 

streaming_and_repair_compaction_mode = {
'disabled',
'data',
'data-and-tombstone'
}

File: test/topology_custom/test_repair.py (hunk 3 above)
Comment (denesb) [reply test/topology_custom/test_repair.py:104]: https://github.com/scylladb/scylladb/pull/17559

File: test/topology_custom/test_repair.py (hunk 4 above)
Comment (denesb) [reply test/topology_custom/test_repair.py:107]: Simple flush left commitlog behind, and tombstone GC wouldn't work, hence I resorted to restarting the node.

**Comment by denesb** [comment]
> Should we make it true when tombstone_gc = repair?

One might also want to enable this for regular timeout, tombstone_gc. The use-case motivating this PR is TWCS table, with `gc_grace_seconds=0` (practically immediate-mode GC).

File: db/config.cc (hunk 1 above)
Comment (denesb) [reply db/config.cc:974]: >  The word "compaction" is misleading here. People will see it as a compaction process that reads some sstables and writes others. But here it's just about eliminating tombstones (and compacting_reader is just half of what users see as compaction).

I can remove the compaction from the name.

> streaming_and_repair_compaction_mode = {
>'disabled',
>'data',
> 'data-and-tombstone'
>}

This flag is already in many releases, we cannot change its type now.

**Comment by denesb** [comment]
> @denesb - who needs to resolve conflicts here?

The author of the PR. I'll tell him to hurry up.

**Comment by denesb** [comment]
V2:
* Rename config item: drop `_compaction` suffix.
* Rephrase the config item description, make the connection to `enable_compacting_data_for_streaming_and_repair` clearer.
* Rebase.

**Comment by raphaelsc** [comment]
why don't we patch sstables_as_snapshot_source() too, to drop expired tombstones with the new "compact_on_read" flag turned one? read repair is another source that can bounce dead tombstones around, right?

File: db/config.cc (hunk 1 above)
Comment (asias) [reply db/config.cc]: It is sad but ok.

**Comment by denesb** [comment]
> why don't we patch sstables_as_snapshot_source() too, to drop expired tombstones with the new "compact_on_read" flag turned one? read repair is another source that can bounce dead tombstones around, right?

No, read-repair already drops all expired tombstones.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping.

**Comment by raphaelsc** [comment]
> > why don't we patch sstables_as_snapshot_source() too, to drop expired tombstones with the new "compact_on_read" flag turned one? read repair is another source that can bounce dead tombstones around, right?
> 
> No, read-repair already drops all expired tombstones.

Hi @denesb today it came to my mind what happens given that tombstone being considered expired is no longer deterministic.

1st example: repair mode, repairs might have different state, and therefore replica r1 might consider a set of tombstones expired, where as r2 has a different set.
2nd example: commitlog segment age can influence the time point before which tombstones are expired, and that's also local state. 

so there could be times in which we bounce expired tombstones around through read repair?

I remember we have plans to make repair history global, and that could help. as for commitlog, I think that will not be a problem after Callee patch that introduces a maximum age for segments.

**Comment by denesb** [comment]
> Hi @denesb today it came to my mind what happens given that tombstone being considered expired is no longer deterministic.
> 
> 1st example: repair mode, repairs might have different state, and therefore replica r1 might consider a set of tombstones expired, where as r2 has a different set. 2nd example: commitlog segment age can influence the time point before which tombstones are expired, and that's also local state.
> 
> so there could be times in which we bounce expired tombstones around through read repair?


Read repair purges tombstones so it won't bounce expired tombstones around. Regular repair does indeed, but with this patch it can be made to also purge tombstones and not do that anymore.

> I remember we have plans to make repair history global, and that could help. as for commitlog, I think that will not be a problem after Callee patch that introduces a maximum age for segments.

**Comment by raphaelsc** [comment]
> > Hi @denesb today it came to my mind what happens given that tombstone being considered expired is no longer deterministic.
> > 1st example: repair mode, repairs might have different state, and therefore replica r1 might consider a set of tombstones expired, where as r2 has a different set. 2nd example: commitlog segment age can influence the time point before which tombstones are expired, and that's also local state.
> > so there could be times in which we bounce expired tombstones around through read repair?
> 
> Read repair purges tombstones so it won't bounce expired tombstones around. Regular repair does indeed, but with this patch it can be made to also purge tombstones and not do that anymore.
> 
> > I remember we have plans to make repair history global, and that could help. as for commitlog, I think that will not be a problem after Callee patch that introduces a maximum age for segments.

what I was trying to say is that what one thinks is expired, the other might not think so. so I say that tombstone A is expired based on my local state. you think tombstone A isn't based on your local state, then on read repair, I will push my expired tombstone to you. what's that local state? repair history table / commitlog segment age. both of them can influence what is the gc_before time point. maybe I am missing something and it's not a problem?

**Comment by raphaelsc** [comment]
actually the other way around in my example, you'll push the tombstone, you don't consider as expired, to me.

**Comment by denesb** [comment]
Compaction happening on the read paths, including the one for repair, do not take other sstables/memtables/commitlog into account when deciding whether a tombstone is garbage collectible. All eligible tombstones are garbage collected. So as long as you repair all replicas (and thus have the same repair history on all replicas), you will not see any differences between nodes.

**Comment by raphaelsc** [comment]
but I think commitlog won't be a problem with Calle's recent work to limit segment age to a constant. so nodes in the same cluster should have more or less the same notion of grace period. right?

**Pull request merged**